apps/tp_dump.py

```diff
@@ -108,15 +111,106 @@ def tp_channel_histogram(tp_data, quiet=False):
+             }
+    labels = {
+                'adc_integral': "ADC Integral",
+                'adc_peak': "ADC Count",
+                'algorithm': "",
+                'channel': "Channel Number",
+                'detid': "",
+                'flag': "",
+                'time_over_threshold': "Ticks",
+                'time_peak': "Ticks",
+                'time_start': "Ticks",
+                'type': "",
+                'version': ""
+             }
+
+    anomaly_filename = 'tp_anomaly_summary.txt'
+
+    if not no_anomaly:
+        if not quiet:
+            print(f"Writing descriptive statistics to {anomaly_filename}.")
+        import os
+        from scipy import stats # Only used when writing to file.
+        if os.path.isfile(anomaly_filename):
+            # Prepare a new tp_anomaly_summary.txt
+            os.remove(anomaly_filename)
+
+    with PdfPages("tp_summary_stats.pdf") as pdf:
+        for tp_key, title in titles.items():
+            # Extract only ta_key
+            plot_data = []
+            for frag_data in tp_data:
+                plot_data = plot_data + list(frag_data[tp_key])
+            plot_data = np.array(plot_data)
+            plt.figure(figsize=(6,4))
+            ax = plt.gca()
+
+            plt.boxplot(plot_data, notch=True, vert=False, sym='+')
+            plt.yticks([])
+            ax.xaxis.grid(True)
+            plt.xlabel(labels[tp_key])
+
+            plt.title(title)
+
+            plt.tight_layout()
+            pdf.savefig()
+            plt.close()
+
+            # Write anomalies to file.
+            if not no_anomaly:
+                if "Sanity" in title and np.all(plot_data == plot_data[0]):
+                    # Either passed check or all wrong in the same way.
+                    continue
+                summary = stats.describe(plot_data)
+                std = np.sqrt(summary.variance)
+                with open(anomaly_filename, 'a') as out:
+                    out.write(f"{title}\n")
+                    out.write(f"Reference Statistics:\n"            \
+                              f"\tTotal # TPs = {summary.nobs},\n"  \
+                              f"\tMean = {summary.mean:.2f},\n"     \
+                              f"\tStd = {std:.2f},\n"               \
+                              f"\tMin = {summary.minmax[0]},\n"     \
+                              f"\tMax = {summary.minmax[1]}.\n")
+                    std3_count = np.sum(plot_data > summary.mean + 3*std) \
+                               + np.sum(plot_data < summary.mean - 3*std)
+                    std2_count = np.sum(plot_data > summary.mean + 2*std) \
+                               + np.sum(plot_data < summary.mean - 2*std)
+                    out.write(f"Anomalies:\n"                           \
+                              f"\t# of >3 Sigma TPs = {std3_count},\n"  \
+                              f"\t# of >2 Sigma TPs = {std2_count}.\n")
+                    out.write("\n\n")
```

Comment:
Maybe not now, but for the future, would be good to abstract it into a separate different function. Maybe even keep it in a different file, so that the ta_dump can re-use this too. I imagine this will be part of the rewriting this into more unified framework in the future.